pkg/runtime-manager/resource-executor/resource_executor.go

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2022 The Koordinator Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package resource_executor
+
+type RuntimeResourceExecutor interface {
+	GenerateResourceCheckpoint() interface{}
+	GenerateHookRequest() interface{}
+	ParseRequest(request interface{}) error
+	ResourceCheckPoint(response interface{}) error
+}
```

Comment:
Is it mainly used for auditing?

Reply:
resource_executor's lifecycle is one grpc/http API process session, in charge of:
1. parse cri/docker request and extract useful message like containerResource needed by protocol between RuntimeHoookServer and RuntimeManager.
2. checkpoint the useful message like containerResource to localdb(bolt). for PreStartContainerHook example,  in CRI StartContainer, there is only one containerID in request, the container related resource info should be stored during CreateContainer stage and indexed by containerID.
...

all these operations are aggregated to interface resource_executor to make the prehook->call backend->posthook process readable:
->resourceExecutor.ParseRequest(request)
 -> ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PreHook, resourceExecutor.GenerateHookRequest())
   -> resourceExecutor.ResourceCheckPoint(res);
     -> ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PostHook, resourceExecutor.GenerateHookRequest())

the runPodsanbox/createcontainer.. api should implement itself resource_executor, and cri version runPodsanbox/createcontainer would be offered in next patch.

Reply:
> resource_executor's lifecycle is one grpc/http API process session, in charge of:
> 
> 1. parse cri/docker request and extract useful message like containerResource needed by protocol between RuntimeHoookServer and RuntimeManager.
> 2. checkpoint the useful message like containerResource to localdb(bolt). for PreStartContainerHook example,  in CRI StartContainer, there is only one containerID in request, the container related resource info should be stored during CreateContainer stage and indexed by containerID.
>    ...
> 
> all these operations are aggregated to interface resource_executor to make the prehook->call backend->posthook process readable: ->resourceExecutor.ParseRequest(request) -> ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PreHook, resourceExecutor.GenerateHookRequest()) -> resourceExecutor.ResourceCheckPoint(res); -> ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PostHook, resourceExecutor.GenerateHookRequest())
> 
> the runPodsanbox/createcontainer.. api should implement itself resource_executor, and cri version runPodsanbox/createcontainer would be offered in next patch.
resource_executor's lifecycle is one grpc/http API process session, in charge of:

parse cri/docker request and extract useful message like containerResource needed by protocol between RuntimeHoookServer and RuntimeManager.
checkpoint the useful message like containerResource to localdb(bolt). for PreStartContainerHook example, in CRI StartContainer, there is only one containerID in request, the container related resource info should be stored during CreateContainer stage and indexed by containerID.
...
all these operations are aggregated to interface resource_executor to make the prehook->call backend->posthook process readable:
->resourceExecutor.ParseRequest(request)
-> ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PreHook, resourceExecutor.GenerateHookRequest())
-> resourceExecutor.ResourceCheckPoint(res);
-> ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PostHook, resourceExecutor.GenerateHookRequest())

the runPodsanbox/createcontainer.. api should implement itself resource_executor, and cri version runPodsanbox/createcontainer would be offered in next patch.

Reply:
> Is it mainly used for auditing?

resource_executor's lifecycle is one grpc/http API process session, in charge of:

1. parse cri/docker request and extract useful message like containerResource needed by protocol between RuntimeHoookServer and RuntimeManager.
2. checkpoint the useful message like containerResource to localdb(bolt). for PreStartContainerHook example, in CRI StartContainer, there is only one containerID in request, the container related resource info should be stored during CreateContainer stage and indexed by containerID.
...
all these operations are aggregated to interface resource_executor to make the prehook->call backend->posthook process readable:
->resourceExecutor.ParseRequest(request)
-> ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PreHook, resourceExecutor.GenerateHookRequest())
-> resourceExecutor.ResourceCheckPoint(res);
-> ci.dispatcher.Dispatch(ctx, runtimeHookPath, config.PostHook, resourceExecutor.GenerateHookRequest())

the runPodsanbox/createcontainer... api should implement itself resource_executor, and cri version runPodsanbox/createcontainer... would be offered in next patch.

Reply:
Is resourceExecutor.GenerateHookRequest()  tracking the modified part of the prehook?

Reply:
> 

a new api would be introduced like resourceExecutor.UpdateResource(response)

Reply:
> > 
> 
> a new api would be introduced like resourceExecutor.UpdateResource(response)

a new api would be introduced like resourceExecutor.UpdateResource(response)

Reply:
> Is resourceExecutor.GenerateHookRequest() tracking the modified part of the prehook?

a new api would be introduced like resourceExecutor.UpdateResource(response)